lib/value/src/value/iter.rs

```diff
@@ -0,0 +1,466 @@
+use crate::Value;
+use std::{marker::PhantomData, ops::IndexMut};
+
+impl Value {
+    /// Create an iterater over the `Value`.
+    ///
+    /// For non-collection types, this returns a single-item iterator similar to
+    /// `Option`'s iterator implementation.
+    ///
+    /// For collection types, it returns all elements in the collection.
+    ///
+    /// The resulting item is an [`IterItem`], which contains either a mutable
+    /// `Value` for non-collection types, a (`&mut String`, `&mut Value`) pair for
+    /// object-type collections, or an immutable/mutable (`usize`, `&mut Value`)
+    /// pair for array-type collections.
+    ///
+    /// ## Recursion
+    ///
+    /// If `recursion` is set to `true`, the iterator recurses into nested
+    /// collection types.
+    ///
+    /// Recursion follows these rules:
+    ///
+    /// - When a collection type is found, that type is first returned as-is.
+    ///   That is, if we're iterating over an object, and within that object is
+    ///   a field "foo" containing an array, then we first return
+    ///   `IterItem::KeyValue`, which contains the key "foo", and the array as
+    ///   the value.
+    /// - After returning the collection type, the iterator recurses into the
+    ///   nested collection itself. Using the previous example, we now go into
+    ///   the array, and start returning `IterItem::IndexValue` variants for the
+    ///   elements within the array.
+    /// - Any mutations done to the array before recursing into it are
+    ///   preserved, meaning once recursion starts, the mutations done on the
+    ///   object itself are preserved.
+    pub fn into_iter<'a>(self, recursive: bool) -> ValueIter<'a> {
+        let data = match self {
+            Value::Object(object) => IterData::Object(object.into_iter().collect()),
+            Value::Array(array) => IterData::Array(array),
+            value => IterData::Value(value),
+        };
+
+        ValueIter::new(data, recursive)
+    }
+}
+
+/// An [`Iterator`] over a [`Value`].
+pub struct ValueIter<'a> {
+    data: IterData,
+    recursive: bool,
```

Comment:
Related to the recursion question above, if there are separate functions, would there be further utility in having separate return types for the two execution paths?

Reply:
There would, because we want to return both index/value and key/value pairs in the same iteration as well.

Reply:
I mean that the non-recursive function would return a `struct ValueIter` with no recursion support, and the recursive function would return a `struct ValueRecursingIter`.